### PR TITLE
[UNet2DConditionModel] add gradient checkpointing

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds gradient checkpoint for the `UNet2DConditionModel`.
It can be enabled or disabled via `enable_gradient_checkpointing ` and `disable_gradient_checkpointing ` methods in `ModelMixin`.

To enable gradient checkpointing:

```python
model = UNet2DConditionModel(...)
model.enable_gradient_checkpointing()
```

### CODE REVIEW

**Comment by patil-suraj** [comment]
This should be ready top merge today :)

File: src/diffusers/modeling_utils.py [hunk 1]
```diff
@@ -124,10 +125,42 @@ class ModelMixin(torch.nn.Module):
...
+    def is_gradient_checkpointing(self) -> bool:
+        """
+        Whether gradient checkpointing is activated for this model or not.
+
+        Note that in other frameworks this feature can be referred to as "activation checkpointing" or "checkpoint
+        activations".
+        """
+        return any(hasattr(m, "gradient_checkpointing") and m.gradient_checkpointing for m in self.modules())
+
+    def gradient_checkpointing_enable(self):
```
Comment (anton-l) [inline src/diffusers/modeling_utils.py:143]:
```suggestion
    def enable_gradient_checkpointing(self):
```

Now that we have the enable_* pattern with `enable_attention_slicing` :sweat_smile:

File: src/diffusers/modeling_utils.py [hunk 2]
```diff
@@ -124,10 +125,42 @@ class ModelMixin(torch.nn.Module):
     """
     config_name = CONFIG_NAME
     _automatically_saved_args = ["_diffusers_version", "_class_name", "_name_or_path"]
+    supports_gradient_checkpointing = False
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:128]: Let's make it maybe private here no? Or do we people to access the variable?

File: src/diffusers/models/unet_2d_condition.py [hunk 3]
```diff
@@ -54,6 +63,8 @@ class UNet2DConditionModel(ModelMixin, ConfigMixin):
         attention_head_dim (`int`, *optional*, defaults to 8): The dimension of the attention heads.
     """
 
+    supports_gradient_checkpointing = True
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:66]:
```suggestion
    _supports_gradient_checkpointing = True
```

File: src/diffusers/modeling_utils.py (hunk 2 above)
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:128]:
```suggestion
    _supports_gradient_checkpointing = False
```

Comment (patrickvonplaten) [reply src/diffusers/modeling_utils.py:128]: (nit)

**Review by patrickvonplaten — APPROVED** [review]
Very nice!

**Review by anton-l — APPROVED** [review]
LGTM!
